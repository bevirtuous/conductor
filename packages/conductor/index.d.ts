import { EventEmitter } from 'events';
import { History } from 'history';

/**
 * An *action* represents the native history action that is performed when
 * manipulating the browser's history.
 */
export type Actions = 'POP' | 'PUSH' | 'REPLACE' | 'RESET' | 'UPDATE';

/**
 * The parameters for the router event callback function.
 */
export interface CallbackParams {
  prev: Route;
  next: Route;
}

/**
 * A function that is called when a router event is emitted.
 */
export type Callback = (p: CallbackParams) => void;

/**
 * Will be triggered when ever the router *will push* a new route.
 * @param callback {Function} The callback function to call once the event has been triggered.
 */
export function onWillPush(callback: Callback): void;

/**
 * Will be triggered when ever the router *did push* a new route.
 * @param callback {Function} The callback function to call once the event has been triggered.
 */
export function onDidPush(callback: Callback): void;

/**
 * Will be triggered when ever the router *will pop* from a route.
 * @param callback {Function} The callback function to call once the event has been triggered.
 */
export function onWillPop(callback: Callback): void;

/**
 * Will be triggered when ever the router *did pop* from a route.
 * @param callback {Function} The callback function to call once the event has been triggered.
 */
export function onDidPop(callback: Callback): void;

/**
 * Will be triggered when ever the router *will replace* a route.
 * @param callback {Function} The callback function to call once the event has been triggered.
 */
export function onWillReplace(callback: Callback): void;

/**
 * Will be triggered when ever the router *did replace* a route.
 * @param callback {Function} The callback function to call once the event has been triggered.
 */
export function onDidReplace(callback: Function): void;

/**
 * Will be triggered when ever the router *will reset* the history.
 * @param callback {Function} The callback function to call once the event has been triggered.
 */
export function onWillReset(callback: Callback): void;

/**
 * Will be triggered when ever the router *did reset* the history.
 * @param callback {Function} The callback function to call once the event has been triggered.
 */
export function onDidReset(callback: Callback): void;

/**
 * Will be triggered when ever the router *updates* a route.
 * @param callback {Function} The callback function to call once the event has been triggered.
 */
export function onUpdate(callback: Callback): void;

/**
 * The event emitter that is used to emit browser events that someone can register for.
 */
export type Emitter = EventEmitter;

/**
 * The ID of a single route.
 */
export type RouteId = string;

/**
 * Object whose values correspond to URL parameters defined in the route definition.
 */
export interface Params {
  [anyProps: string]: any
}

/**
 * Object whose values correspond to the URL query string individuals.
 */
export interface Query {
  [anyProps: string]: any
}

/**
 * Object containing custom key value pairs. This can be used to include
 * arbitrary data when opening a route.
 */
export interface State {
  [anyProps: string]: any
}

/**
 * The parameters a single route definition can receive.
 */
export interface RouteParams {
  id: RouteId;
  pathname: string;
  pattern?: string;
  state?: object;
  transform?: Function;
}

/**
 * The instance of a *route* that the router will create.
 */
export class Route {
  constructor(params: RouteParams)
  id: RouteId;
  pathname: string;
  pattern: string;
  location: string;
  params: Params;
  query: Query;
  hash: string;
  state: State;
  created: number;
  updated: number | null;
}

/**
 * The *stack* holds all active routes.
 */
export class Stack {
  add: (id: RouteId, entry: Route) => void;
  first: () => void;
  get: (id: RouteId) => Route;
  getByIndex: (index: number | null) => Route | null;
  getAll: () => Map<RouteId, Route>;
  last: () => Route | null;
  clear: () => void;
  remove: (id: string) => void;
  reset: (p: Array<RouteId, Route>) => void;
  update: (id: string, entry: Route) => void;
}

/**
 * The parameters for every route *actions* that can be emitted.
 */
export type ActionParams = {
  emit: boolean;
  steps: Number;
  state: Object | null;
}

/**
 * The *router* handles all the logic when navigating and controls the individual routes.
 */
export class Router {
  constructor(createHistory: History);
  nativeEvent: boolean;
  history: History;
  patterns: object;
  routing: boolean;
  routeIndex: number;
  action: Actions;
  handleNativeEvent: (location: Route, action: Actions) => void;
  createId: () => void;
  addInitialRoute: () => void;
  handlePop: (params: ActionParams) => Promise<object>;
  handlePush: (params: ActionParams, override?: boolean) => Promise;
  findPattern: (pathname: string) => string | null;
  register: (pattern: string, transform: Function | null) => void;
  handleReplace: (params: ActionParams) => Promise;
  push: (params: ActionParams) => Promise;
  pop: (params: ActionParams) => Promise;
  replace: (params: ActionParams) => Promise;
  reset: () => Promise;
  resetTo: (pathname: string) => Promise;
  update: (id: string, state?: IState, emit?: boolean) => Promise;
  getCurrentRoute: () => Route;
}
