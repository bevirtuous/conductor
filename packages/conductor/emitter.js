module.exports = require('./src/emitter');
