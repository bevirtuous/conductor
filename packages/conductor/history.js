module.exports = require('./src/history');
