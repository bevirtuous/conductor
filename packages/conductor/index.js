module.exports = require('./src/Conductor');
