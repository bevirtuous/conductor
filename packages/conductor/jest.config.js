const defaultConfig = require('@virtuous/react-unit-test-suite/jest.config');

const config = {
  ...defaultConfig,
  testURL: 'http://localhost',
};

module.exports = config;
