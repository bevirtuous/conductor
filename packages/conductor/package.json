{
  "name": "@virtuous/conductor",
  "version": "2.0.11",
  "description": "A javascript router",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "git+ssh://git@github.com:bevirtuous/conductor.git"
  },
  "bugs": {
    "url": "https://github.com/bevirtuous/conductor/issues"
  },
  "homepage": "https://github.com/bevirtuous/conductor/tree/master/packages/conductor",
  "keywords": [
    "conductor",
    "javascript",
    "react",
    "redux",
    "router",
    "route",
    "routing",
    "virtuous",
    "history"
  ],
  "contributors": [
    "Richard Gorman <richardgorman@outlook.com> (https://github.com/richardgorman)",
    "Florian MÃ¼ller <flomueller1981@gmail.com> (https://github.com/devbucket)"
  ],
  "main": "./index.js",
  "typings": "./index.d.ts",
  "scripts": {
    "test": "NODE_ENV=test npx jest",
    "cover": "NODE_ENV=test npx nyc jest --coverage"
  },
  "dependencies": {
    "history": "^4.7.2",
    "path-match": "^1.2.4",
    "query-string": "^6.1.0",
    "url-pattern": "^1.0.3"
  },
  "nyc": {
    "include": [
      "src/**/*.js"
    ],
    "exclude": [
      "build",
      "test",
      "dist",
      ".nyc_output",
      "node_modules"
    ],
    "reporter": [
      "lcov",
      "text-summary"
    ],
    "all": true,
    "report-dir": "build"
  }
}
