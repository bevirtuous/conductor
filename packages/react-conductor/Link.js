module.exports = require('./src/Link');
