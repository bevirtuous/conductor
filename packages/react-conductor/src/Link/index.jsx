import React from 'react';
import PropTypes from 'prop-types';
import conductor from '@virtuous/conductor';
import {
  ACTION_PUSH,
  ACTION_POP,
  ACTION_REPLACE,
} from '@virtuous/conductor/constants';

/**
 * The Link component.
 */
class Link extends React.Component {
  static propTypes = {
    children: PropTypes.element.isRequired,
    to: PropTypes.string.isRequired,
    action: PropTypes.oneOf([ACTION_PUSH, ACTION_POP, ACTION_REPLACE]),
    state: PropTypes.shape(),
    tag: PropTypes.string,
  };

  static defaultProps = {
    action: ACTION_PUSH,
    state: {},
    tag: 'a',
  }

  /**
   * @param {MouseEvent} event The click event.
   */
  handleClick = (event) => {
    event.preventDefault();

    const { action, state, to } = this.props;

    if (action === ACTION_PUSH) {
      conductor.push(to, state);
    } else if (action === ACTION_POP) {
      conductor.pop();
    } else {
      conductor.replace(to, state);
    }
  }

  /**
   * @returns {JSX}
   */
  render() {
    const { children, tag: Component } = this.props;
    return (
      <Component onClick={this.handleClick}>
        {children}
      </Component>
    );
  }
}

export default Link;
