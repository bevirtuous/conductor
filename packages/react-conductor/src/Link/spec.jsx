import React from 'react';
import { shallow } from 'enzyme';
import conductor from '../../../conductor';
import Link from './index';

describe.skip('<Link />', () => {
  beforeAll(() => {
    conductor.register('/mypage');
  });

  it('should render correctly with given props', () => {
    const wrapper = shallow((
      <Link href="/mypage">
        My Link
      </Link>
    ));

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('a').length).toBe(1);
    expect(wrapper.text()).toEqual('My Link');
  });

  it('should push when clicked with default action prop', () => {
    const wrapper = shallow((
      <Link href="/mypage">
        My Link
      </Link>
    ));

    wrapper.simulate('click', { preventDefault() {} });

    expect(wrapper).toMatchSnapshot();
    expect(conductor.stack.length).toBe(1);
  });

  it('should pop when clicked with prop action=REPLACE', () => {
    const wrapper = shallow((
      <Link href="/mypage" action="REPLACE">
        My Link
      </Link>
    ));

    wrapper.simulate('click', { preventDefault() { } });

    expect(wrapper).toMatchSnapshot();
    expect(conductor.stack.length).toBe(1);
  });

  it('should pop when clicked with prop action=POP', () => {
    const wrapper = shallow((
      <Link href="/mypage" action="POP">
        My Link
      </Link>
    ));

    wrapper.simulate('click', { preventDefault() { } });

    expect(wrapper).toMatchSnapshot();
    expect(conductor.stack.length).toBe(0);
  });
});
