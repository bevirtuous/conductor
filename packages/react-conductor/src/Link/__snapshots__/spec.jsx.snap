// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Link /> should pop when clicked with prop action=POP 1`] = `
<a
  onClick={[Function]}
>
  My Link
</a>
`;

exports[`<Link /> should pop when clicked with prop action=REPLACE 1`] = `
<a
  onClick={[Function]}
>
  My Link
</a>
`;

exports[`<Link /> should push when clicked with default action prop 1`] = `
<a
  onClick={[Function]}
>
  My Link
</a>
`;

exports[`<Link /> should render correctly with given props 1`] = `
<a
  onClick={[Function]}
>
  My Link
</a>
`;
